airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/s3/S3CopierSupplierDelegate.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.destination.jdbc.copy.s3;
+
+import com.amazonaws.services.s3.AmazonS3;
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.integrations.destination.ExtendedNameTransformer;
+import io.airbyte.integrations.destination.jdbc.SqlOperations;
+import io.airbyte.integrations.destination.jdbc.copy.Copier;
+import io.airbyte.protocol.models.DestinationSyncMode;
+import java.io.IOException;
+
+public interface S3CopierSupplierDelegate {
```

Comment:
I'm having trouble wrapping my head around the delegation mechanism here. I think I understand why it works but it's a bit hard for me to follow. Here are the parts that are confusing:
1. I'm not able to parse how this is different from the CopierSupplier iface.
2. I'm not fully understanding why `S3CopierSupplier` (which is really a copier factory) is just calling another thing that is just a copier factory?

Would it make sense to just make S3Copier an abstract class and skip the delegation?

Reply:
It's possible this is just a naming thing. Maybe if the `S3CopierSupplierDelegate` iface were called `S3CopyQuery` I wouldn't be confused. I think passing in a query via an interface makes sense to me just fine. The generic name of delegate means I'm pretty unsure what's being passed and why.